File: captainhcg/go/2-add-two-numbers.go
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func addTwoNumbers(l1 *ListNode, l2 *ListNode) *ListNode {
    carry := false
    header := &ListNode{}
    last := header
    for l1 != nil || l2 != nil || carry {
        cur := 0;
        if carry {
            cur += 1 
        }
        if l1 != nil {
            cur += l1.Val
            l1 = l1.Next
        }
        if l2 != nil {
            cur += l2.Val
            l2 = l2.Next
        }
        if cur >= 10 {
            carry = true
            cur -= 10
        } else {
            carry = false
        }
        last.Next = &ListNode{cur, nil}
        last = last.Next
    }
    return header.Next
}

File: captainhcg/js/2-add-two-numbers.js
var addTwoNumbers = function(l1, l2) {
    var carry = 0;
    var head = new ListNode(-1);
    var pre = head;
    while(l1 || l2 || carry){
        var s = carry;
        if(l1){
            s += l1.val;
            l1 = l1.next;
        }
        if(l2){
            s += l2.val;
            l2 = l2.next;
        }
        if(s >= 10){
            s -= 10;
            carry = 1;
        } else {
            carry = 0;
        }
        pre.next = new ListNode(s);
        pre = pre.next;
    }
    if(!head.next){
        head.next = new ListNode(0)
    }
    return head.next;
};
